Vertex color layer and nodes

3DSinghVFX opened the pull request:
1) This branch has Vertex Color Layer as mesh property like UVs. 
2) Two new nodes, **Get Mesh Vertex Color** and **Add Mesh Vertex Color**.

**Get Mesh Vertex Color**: This node takes mesh and the name of the vertex color layer as inputs. It gives the colors based on the color-mode (Loop, Vertex, Polygon).
![Screenshot from 2019-12-18 22-56-35](https://user-images.githubusercontent.com/30294746/71108756-c4bc6900-21e9-11ea-8987-f9fd23d80a0f.png)

**Add Mesh Vertex Color**: This node allows you to add a vertex color layer to the mesh. This node takes mesh, vertex color layer name, and color/colors (based on the color-mode e.g., Loop, Vertex, Polygon) as inputs. 
![Screenshot from 2019-12-18 22-57-02](https://user-images.githubusercontent.com/30294746/71108771-cbe37700-21e9-11ea-9005-d2c2fe3bd8cf.png)

3) The Mesh Info, Mesh Object Input, and Mesh Object Output nodes are updated.

OmarEmaraDev commented on animation_nodes/nodes/mesh/add_mesh_vertex_color.py:
```diff
@@ -0,0 +1,69 @@
+    ("POLYGON", "Polygon", "Get color of every polygon", "NONE", 2)    
+]
+
+class AddMeshVertexColorNode(bpy.types.Node, AnimationNode):
```
Perhaps we should change the name to `InsertVertexColorLayerNode`, to be more uniform and to make it easier to search for. Do you agree?

3DSinghVFX replied:
Omar, `Add Mesh Vertex Color` and  `Get Mesh Vertex Color` nodes do similar jobs, i.e., one node insert the vertex color layer and its colors, and the other gets the colors of a vertex color layer. I think if their names are similar then it would be nice. 

Add Mesh Vertex Color -> Insert Vertex Color Layer
Get Mesh Vertex Color -> Get Vertex Color Layer 

or

Add Mesh Vertex Color -> Mesh Vertex Color Output  or  Vertex Color Layer Output
Get Mesh Vertex Color -> Mesh Vertex Color Input  or  Vertex Color Layer Input

OmarEmaraDev replied:
Yes, sure. This looks good:

Add Mesh Vertex Color -> Insert Vertex Color Layer
Get Mesh Vertex Color -> Get Vertex Color Layer

3DSinghVFX replied:
Okay.

OmarEmaraDev commented on animation_nodes/nodes/mesh/add_mesh_vertex_color.py:
```diff
@@ -0,0 +1,69 @@
+            return "execute_SingleColor"
+
+    def execute_SingleColor(self, mesh, colorLayerName, color):
+        if mesh is None: return
```
return None

3DSinghVFX replied:
Added.

OmarEmaraDev commented on animation_nodes/nodes/mesh/add_mesh_vertex_color.py:
```diff
@@ -0,0 +1,69 @@
+        
+    def execute_ColorsList(self, mesh, colorLayerName, colors):
+        if mesh is None: return None
+        if colorLayerName == "": return mesh
```
We should probably inform the user using exception error handling.

3DSinghVFX replied:
I have added the exception error handling.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_mesh_vertex_color.py:
```diff
@@ -0,0 +1,56 @@
+        layout.prop(self, "colorMode", text = "")
+
+    def execute(self, mesh, colorLayerName):
+        if mesh is None or colorLayerName not in mesh.getVertexColorLayerNames(): return ColorList()
```
Inform the user using exception error handling here as well.

3DSinghVFX replied:
Here, it is also added the exception error handling.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_mesh_vertex_color.py:
```diff
@@ -0,0 +1,56 @@
+        for name, data in mesh.getVertexColorLayers():
+            if name == colorLayerName:
+                colorsList = data
+
```
Add a method to the mesh that returns the vertex color layer for the given name.

3DSinghVFX replied:
Now this is a method which gives vertex color layer for the given name.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -141,13 +154,18 @@ cdef class Mesh:
                 Vertices: {}
                 Edges: {}
```
Try replacing this with an f-string.

3DSinghVFX replied:
Omar, I have checked with the fresh AN compile build, there is no `runtime` warning. It is working fine.

OmarEmaraDev replied:
Good. To clarify, I wanted you to using python's f-strings here. See https://www.python.org/dev/peps/pep-0498/

3DSinghVFX commented:
Omar, I'll remove the Vertex Color Input node. The reason I kept this node is because there is a bug in `color/c_utils.pyx`, so the Vertex Color Input node gives different colors,
![RK1HfKVU jpg large](https://user-images.githubusercontent.com/30294746/71161565-95503f80-226f-11ea-98ef-b5dbabf172fd.jpeg)


https://github.com/JacquesLucke/animation_nodes/blob/af8354fffcb9a05290fe81cebd3ad9ccbcf0c8d7/animation_nodes/nodes/color/c_utils.pyx#L64

When I change this to `addColor_Inplace(polygonColors.data + i, colors.get(index))` then Vertex Color Input node gives the same colors.

OmarEmaraDev commented:
Sure, you may fix the bug in this pull request as well.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -173,6 +191,13 @@ cdef class Mesh:
+            if self.getVertexColorLayers()[i][0] == colorLayerName:
+                return self.getVertexColorLayers()[i]    
+            else:
+                return None
```
`vertexColorLayers` is an ordered dictionary, you can simply get the vertex color layer using subscript notation. Also, move this function below the `getVertexColorLayerNames`.

3DSinghVFX replied:
Is this correct way,
```
def getVertexColors(self, str colorLayerName):
        return self.vertexColorLayers.get(colorLayerName, None)
```
This function return the colors corresponds to the colorLayerName.

OmarEmaraDev replied:
Yes, looks good.

3DSinghVFX replied:
I have added this function.

3DSinghVFX commented:
Omar, I have used the f-strings, here is the code:
```
def __repr__(self):
        return textwrap.dedent(
        f"""AN Mesh Object:
        Vertices: {len(self.vertices)}
        Edges: {len(self.edges)}
        Polygons: {len(self.polygons)}
        UV Maps: {list(self.uvMaps.keys())}
        Vertex Color Layers: {list(self.vertexColorLayers.keys())}""")
```
I think we should change the `AN Mesh Object:` to `AN Mesh:`, is it not?.

OmarEmaraDev commented:
I think *Object* here denotes an instance of the AN Mesh class, not a Blender object. So just leave it for now.

3DSinghVFX commented:
> I think _Object_ here denotes an instance of the AN Mesh class, not a Blender object. So just leave it for now.

Okay.

3DSinghVFX commented:
Omar, In Insert Vertex Color Layer node, if we add a color layer of the same name which already there then does not notify the user. Should we add the exception for that,
![screen](https://user-images.githubusercontent.com/30294746/71254743-ec2e4580-2351-11ea-80d8-1b075219988e.png)
Or the viewer node is enough for user to see whether that color layer is already there or not.

OmarEmaraDev commented:
I think we should do that. Raise an exception informing the user about the existing layer.

I will continue the review on Sunday as I don't have much time right now.

3DSinghVFX commented:
Okay.

No problem :)

3DSinghVFX commented:
Omar, I have noticed another problem with the order of color layers arises if we use Set Vertex Color with Color Layer Index option.

In this example, I have inserted two color layers. Then try setting the color to color layer _AN-Col2_ which has index zero with the Set Vertex Color node, 
![Vertex_Color_Layer3](https://user-images.githubusercontent.com/30294746/71304123-4843a900-23e8-11ea-9d14-314623db95db.png)

However, in this case, if include the existing color layer (_Col_) or insert a color layer before the _AN-Col2_ then the Index get change for _AN-Col2_ color layer for Set Vertex Color node,
![Vertex_Color_Layer4](https://user-images.githubusercontent.com/30294746/71304384-d3726e00-23eb-11ea-9364-064048861680.png)

So, I think we should also use the name of the color layer in the Set Vertex Color node to resolve this problem. Do have any other idea?

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -136,18 +152,21 @@ cdef class Mesh:
+        Edges: {len(self.edges)}
+        Polygons: {len(self.polygons)}
+        UV Maps: {list(self.uvMaps.keys())}
+        Vertex Color Layers: {list(self.vertexColorLayers.keys())}""")
```
Can't we use `getVertexColorLayerNames` here?

3DSinghVFX replied:
Yes. I have changed it, `Vertex Colors: {self.getVertexColorLayerNames()}`.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+class GetVertexColorLayerNode(bpy.types.Node, AnimationNode):
+    bl_idname = "an_GetVertexColorLayerNode"
+    bl_label = "Get Vertex Color Layer"
+    bl_width_default = 155
```
I would rather we don't use custom widths here.

3DSinghVFX replied:
Removed.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+        layout.prop(self, "colorMode", text = "")
+
+    def execute(self, mesh, colorLayerName):
+        if mesh is None: return ColorList()
```
We can use an exception here as well.

3DSinghVFX replied:
Omar, For input mesh,
https://github.com/JacquesLucke/animation_nodes/blob/419e1e73431b17cb5cfa6cce76acc588321c5092/animation_nodes/nodes/mesh/get_vertex_color_layer.py#L24
due to this, I think mesh can't none, is it not?

OmarEmaraDev replied:
An input mesh can be None. But if it is None, I would consider this an issue with the node the produced the None value. So maybe you are right, the None check is not needed here.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+    def execute(self, mesh, colorLayerName):
+        if mesh is None: return ColorList()
+        if colorLayerName == "":
+            self.raiseErrorMessage("No Vertex Color Layer Name.")
```
-> Vertex color layer name can't be empty.

3DSinghVFX replied:
It is replaced.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+        colorsList = ColorList(length = len(mesh.polygons.indices))
+        
+        if mesh.getVertexColors(colorLayerName) is not None:
+            colorsList = mesh.getVertexColors(colorLayerName)
```
You are calling `mesh.getVertexColors` two times here.

3DSinghVFX replied:
It is changed.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+            colorsList = mesh.getVertexColors(colorLayerName)
+        else:
+            self.raiseErrorMessage(f"No Vertex Color Layer of Name '{colorLayerName}'."
+                                   f" Layers: {', '.join(mesh.getVertexColorLayerNames())}")
```
This might be a little long for an error message. Also, `of name` -> `with name`.

3DSinghVFX replied:
I added the `f" Layers: {', '.join(mesh.getVertexColorLayerNames())}"` to avoid guesswork or to check names with Viewer Node. Should I remove this line to make the error message short?

OmarEmaraDev replied:
Yes. The Mesh repr method already provide those names.

3DSinghVFX replied:
Now it looks like this `self.raiseErrorMessage(f"Mesh doesn't has '{colorLayerName}' Vertex Color Layer.")`. is it fine?

OmarEmaraDev replied:
I think this is better:

> Mesh doesn't have a vertex color layer with the name '{colorLayerName}'.

3DSinghVFX replied:
Okay, here it is
![Screenshot from 2019-12-23 00-24-14](https://user-images.githubusercontent.com/30294746/71325992-a5d81280-251a-11ea-8365-80091b7312fd.png)

OmarEmaraDev replied:
Looks good.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,60 @@
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getPolygonColorsFromLoopColors(polygonIndices, colorsList)
+        
+    
```
New line at the end of file.

3DSinghVFX replied:
Removed the extra line.

3DSinghVFX commented:
Omar, How we are going to handle this problem: https://github.com/JacquesLucke/animation_nodes/pull/1256#issuecomment-568158138

OmarEmaraDev commented:
Well, the user can always identify the vertex color by name by changing the mode from the advanced node settings. So I don't really consider this an options.

3DSinghVFX commented:
Okay.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,57 @@
+        if getColorList is not None:
+            colorsList = getColorList
+        else:
+            self.raiseErrorMessage(f"Mesh doesn't have a vertex color layer with the name '{colorLayerName}'.")
```
```
colorsList = mesh.getVertexColors(colorLayerName)
if colorsList is None:
    raise
continue as usual without an else.
```

3DSinghVFX replied:
It is done.

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,57 @@
+
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getPolygonColorsFromLoopColors(polygonIndices, colorsList)
+        
```
Setup your editor to add a new line at the end of file.

3DSinghVFX replied:
Sorry. I'm using Visual Studio.

3DSinghVFX replied:
I have setup a new line in my Visual Studio, this how it shows the code,
![Screenshot from 2019-12-23 01-56-05](https://user-images.githubusercontent.com/30294746/71326936-a6c37100-2527-11ea-9025-f50dd79e88ab.png)

OmarEmaraDev commented on animation_nodes/nodes/mesh/get_vertex_color_layer.py:
```diff
@@ -0,0 +1,54 @@
+            self.raiseErrorMessage("Vertex color layer name can't be empty.")
+        
+        defaultColor = Color((0, 0, 0, 1))
+        colorsList = ColorList(length = len(mesh.polygons.indices))
```
This is redundant.

3DSinghVFX replied:
Removed.

OmarEmaraDev commented on animation_nodes/nodes/mesh/mesh_object_input.py:
```diff
@@ -6,6 +6,7 @@
 class MeshObjectInputNode(bpy.types.Node, AnimationNode):
     bl_idname = "an_MeshObjectInputNode"
     bl_label = "Mesh Object Input"
+    bl_width_default = 145
```
I would prefer if we keep the default width here as well.

3DSinghVFX replied:
Now it has the default width.

OmarEmaraDev commented on animation_nodes/nodes/mesh/mesh_object_input.py:
```diff
@@ -14,6 +15,7 @@ def create(self):
         self.newInput("Boolean", "Use World Space", "useWorldSpace")
         self.newInput("Boolean", "Use Modifiers", "useModifiers", value = False)
         self.newInput("Boolean", "Load UVs", "loadUVs", value = False)
+        self.newInput("Boolean", "Load Vertex Colors", "loadVertexColors", value = False)
```
Can you hide both of those inputs by default?

3DSinghVFX replied:
Done.

3DSinghVFX commented on animation_nodes/nodes/mesh/mesh_object_input.py:
```diff
@@ -33,7 +32,7 @@ def create(self):
         self.newOutput("Integer List", "Material Indices", "materialIndices")
 
         self.newOutput("Text", "Mesh Name", "meshName")
```
Omar, Mesh name output has a bug.

OmarEmaraDev replied:
What is the bug?

3DSinghVFX replied:
When I connect the viewer node to the mesh name output then it breaks the node-tree:
![Screenshot from 2019-12-23 15-45-23](https://user-images.githubusercontent.com/30294746/71352107-67cf0300-259b-11ea-8f68-62e43fd068fc.png)

OmarEmaraDev replied:
Fixed in b442dfe.

3DSinghVFX replied:
Okay :)

OmarEmaraDev commented:
Can you add the changes to the CHANGELOG file? Notice that we are still on the same sub version, so adjust the file appropriately. The node that was removed in this pull request should be removed from the CHANGELOG as well.

3DSinghVFX commented:
Okay, I'll update the CHANGELOG file.

3DSinghVFX commented:
I have added the changes about the insert and get vertex color layer nodes, and removed the `- Added Vertex Color Input node.`. Do I have to add anything else?

```
## 2.1.5

### Added

- Added *Include End Point* option to the *Float Range* node.
- Added collection option to *Execution Triggers*.
- Added *Sound Falloff* node.
- Added *Subprograms* to the search menu.
- Added missing nodes to the node menu
- Added *Collection Operations* node.
- Added *Rotations* output to the *Particles Data* node.
- Added Get Vertex Color Layer node.
- Added Insert Vertex Color Layer node.
```

OmarEmaraDev commented:
No, look good.

OmarEmaraDev commented on CHANGELOG.md:
```diff
@@ -37,6 +36,8 @@
 - Added *Collection Operations* node.
 - Added *Rotations* output to the *Particles Data* node.
+- Added Get Vertex Color Layer node.
+- Added Insert Vertex Color Layer node.
```
Why did you add them to 2.1.5? They should go to the Unreleased section.

3DSinghVFX replied:
Sorry :) Now they are in the Unreleased section under the Added category.

OmarEmaraDev commented:
Shouldn't the Insert Vertex Color Layer node have options for Loop, Vertex, and Polygon like the Set Vertex Color node?

OmarEmaraDev commented:
My bad, just realized the options are there.

3DSinghVFX commented:
Thank you for helping me to understand the different aspects of AN :)
